keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/ApplicationService.kt

```diff
@@ -102,18 +103,17 @@ class ApplicationService(
     // TODO: publish ArtifactUnpinnedEvent
   }
 
-  fun markAsVetoedIn(application: String, veto: EnvironmentArtifactVeto, force: Boolean) {
+  fun markAsVetoedIn(user: String, application: String, veto: EnvironmentArtifactVeto, force: Boolean) {
     val config = repository.getDeliveryConfigForApplication(application)
-    val artifact = config.matchingArtifactByReference(veto.reference)
-      ?: throw ArtifactNotFoundException(veto.reference, config.name)
-
-    repository.markAsVetoedIn(
+    val succeeded = repository.markAsVetoedIn(
       deliveryConfig = config,
-      artifact = artifact,
-      version = veto.version,
-      targetEnvironment = veto.targetEnvironment,
+      veto = veto.copy(vetoedBy = user),
       force = force
     )
+    if (!succeeded) {
+      throw InvalidVetoException(application, veto.targetEnvironment, veto.reference, veto.version)
+    }
+    // TODO: publish ArtifactVetoedEvent
```

Comment:
Is this `TODO` something that should be done for this PR, or can it come later?

Reply:
no, it's for later. the best kind of todo :D 

We need to surface events other than resource events, and we don't do that now. So I don't want to add any events